benchmarks/bm_runner.py

```diff
@@ -0,0 +1,371 @@
+
+    def __init__(self, subparsers: ArgumentParser.add_subparsers) -> None:
+        self.subparser: ArgumentParser = subparsers.add_parser(
+            self.name,
+            description=self.description,
+            formatter_class=argparse.RawTextHelpFormatter,
+        )
+        self.add_arguments()
+        self.subparser.add_argument(
+            "asv_args",
+            nargs=argparse.REMAINDER,
+            help="Any number of arguments to pass down to ASV.",
+        )
+        self.subparser.set_defaults(func=self.func)
+
+    @abstractmethod
+    def add_arguments(self) -> None:
+        """All self.subparser.add_argument() calls."""
+        _ = NotImplemented
+
+    @staticmethod
+    @abstractmethod
+    def func(args: argparse.Namespace):
+        """
+        The function to return when the subparser is parsed.
+
+        `func` is then called, performing the user's selected sub-command.
+
+        """
+        _ = args
+        return NotImplemented
+
+
+class Overnight(_SubParserGenerator):
+    name = "overnight"
+    description = (
+        "Benchmarks all commits between the input **first_commit** to ``HEAD``, "
+        "comparing each to its parent for performance shifts. If a commit causes "
+        "shifts, the output is saved to a file:\n"
+        "``.asv/performance-shifts/<commit-sha>``\n\n"
+        "Designed for checking the previous 24 hours' commits, typically in a "
+        "scheduled script."
+    )
+
+    def add_arguments(self) -> None:
+        self.subparser.add_argument(
+            "first_commit",
+            type=str,
+            help="The first commit in the benchmarking commit sequence.",
+        )
+
+    @staticmethod
+    def func(args: argparse.Namespace) -> None:
+        _setup_common()
+
+        commit_range = f"{args.first_commit}^^.."
+        asv_command = ASV_HARNESS.format(posargs=commit_range)
+        subprocess.run([*asv_command.split(" "), *args.asv_args])
+
+        # git rev-list --first-parent is the command ASV uses.
+        git_command = f"git rev-list --first-parent {commit_range}"
+        commit_string = subprocess.run(
+            git_command.split(" "), capture_output=True, text=True
+        ).stdout
+        commit_list = commit_string.rstrip().split("\n")
+        _asv_compare(*reversed(commit_list), overnight_mode=True)
+
+
+class Branch(_SubParserGenerator):
+    name = "branch"
+    description = (
+        "Performs the same operations as ``overnight``, but always on two commits "
+        "only - ``HEAD``, and ``HEAD``'s merge-base with the input "
+        "**base_branch**. Output from this run is never saved to a file. Designed "
+        "for testing if the active branch's changes cause performance shifts - "
+        "anticipating what would be caught by ``overnight`` once merged.\n\n"
+        "**For maximum accuracy, avoid using the machine that is running this "
+        "session. Run time could be >1 hour for the full benchmark suite.**"
```

Comment:
I would find it helpful to include examples for these subcommands. Either in the help for the sub command or maybe in the readme?

Reply:
6322d86599ed44424544105a27acf555d95d9fce